Add dereference blog post (#70)
Apicurio/apicurio.github.io: _posts/2024-05-14-registry-dereference.md

---
layout: post
title: "Apicurio Registry - JSON Schema dereference"
date:   2024-05-14 16:00:00
author: carles
categories: registry openapi jsonchema asyncapi avro
---

As it can be seen in the [documentation](https://www.apicur.io/registry/docs/apicurio-registry/2.5.x/getting-started/assembly-managing-registry-artifacts-api.html), Apicurio Registry allows to manage artifacts with references.

> One of the cool features we've added on top of this is the possibility of using a dereference parameter for certain API endpoints that, when used, allows you to fetch the full content of an artifact with all the references inlined within the same content.

Reviewer: ```suggestion
One of the cool features we've added on top of this is the possibility of using a dereference parameter for certain API endpoints that optionally allows you to fetch the full content of an artifact with all the references inlined within the same content.
```

> This is especially useful in certain contexts to reduce the number of HTTP requests, like, as we will see in this blog, in the Kafka Serializers and Deserializers.

Reviewer: ```suggestion
This is especially useful in certain contexts to reduce the number of HTTP requests in the Kafka Serializers and Deserializers, as you will see in this blog.
```

---

Intro
===

The code example used for this blog post can be found in the Apicurio Registry [examples](https://github.com/Apicurio/apicurio-registry/blob/2.6.x/examples/serdes-with-references/src/main/java/io/apicurio/registry/examples/references/JsonSerdeReferencesDereferencedExample.java).

> One of the limitations of the JSON Schema format is that, given some data that adheres to a particular schema it does not allow to discover the full schema from that content. That's the main reason why, for this serde, the schema being used has to be registered upfront, and that's what you see

Reviewer: ```suggestion
One of the limitations of the JSON Schema format is that you cannot discover the full schema from some data that adheres to the schema. That's the main reason why, for this serde, the schema being used has to be registered upfront, and that's what you see
```

> in the code example from line [57](https://github.com/Apicurio/apicurio-registry/blob/2.6.x/examples/serdes-with-references/src/main/java/io/apicurio/registry/examples/references/JsonSerdeReferencesDereferencedExample.java#L57) to line [132](https://github.com/Apicurio/apicurio-registry/blob/2.6.x/examples/serdes-with-references/src/main/java/io/apicurio/registry/examples/references/JsonSerdeReferencesDereferencedExample.java#L132).

Reviewer 2: Why not a single link to the code, like this:

https://github.com/Apicurio/apicurio-registry/blob/2.6.x/examples/serdes-with-references/src/main/java/io/apicurio/registry/examples/references/JsonSerdeReferencesDereferencedExample.java#L57-L132

Author: I thought there was a reason, but re-reading this there isn't, so changing it.

> Where the main schema _city.json_ is registered with all the references in the hierarchy. Along those lines, a full hierarchy of schemas is registered in Apicurio Registry. The hierarchy can be described as follows and represented as a tree:

Reviewer: Not sure what is intended here:
Where the main schema _city.json_ is registered with all the references in the hierarchy.
maybe
where the main schema _city.json_ is registered with all the references in the hierarchy.

Author: ah, yes, the capital letter should be removed here

```

>                           cityzen

Reviewer: ```suggestion
                          citizen
```

     |          |                     |                   |
    city   qualification        sample.address     citizenIdentifier
     |                                                     |
 qualification                                       qualification
    
```

> As you may notice, there are three different files with the same name. This is done on purpose to demonstrate that you can have different schemas with the same name within the same hierarchy.

Reviewer: I didn't know they were file names, should I have known?

Author: Not necessarily. They might be URLs to other schemas, they just happen to be files in this example.

When they're registered in the code example, they're registered under the same group but using different artifact ids, so they can be referenced from their main schema.

Since an image is way better than trying to describe it, this is the structure in the file system:

> ![Registry (before)](/images/posts/registry-dereference/file-system.png)

Reviewer: maybe this should have been first?

Author: I don't think so. In this case it's just for clarification purposes.

          "description" : "The zip code.",
          "type" : "integer",
          "minimum" : 0
        },
        "qualification" : {
          "title" : "Qualification",
          "type" : "object",
          "properties" : {
            "qualification" : {
              "description" : "The city qualification",
              "type" : "integer",
              "minimum" : 10
            },
            "name" : {
              "description" : "The subject's name",
              "type" : "string"
            }
          }
        },
        "name" : {
          "description" : "The city's name.",
          "type" : "string"
        }
      }
    },
    "age" : {
      "description" : "Age in years which must be equal to or greater than zero.",
      "type" : "integer",
      "minimum" : 0
    }
  },
  "required" : [ "city" ],
  "$id" : "https://example.com/citizen.json"
}
```
As you can see, that's the content of the full hierarchy, not just the schema of the _citizen_ element alone. There are two endpoints where you can get the content dereferenced, with the following curl commands:

```
curl --location 'http://localhost:8080/apis/registry/v2/ids/globalIds/7?dereference=true'

> curl --location 'http://localhost:8080/apis/registry/v2/groups/default/artifacts/JsonSerdeReferencesExample?dereference=true'

Reviewer 2: Do we not support `dereference` on the version specific endpoint?

`/groups/:groupId/artifacts/:artifactId/versions/:versionId`

Author: We do, it's a miss in the blog, thanks!

```

JSON Schema Kafka Serde dereference
===

> Now, in order to make good use of all of this, we can use a Kafka producer. For this example, in the Kafka Producer there is one extremely important property we must set, the schema resolver configuration option _SERIALIZER_DEREFERENCE_SCHEMA_. This is the option what will instruct the Schema Resolver the use the dereference parameter in the globalIds endpoint above (since that is the endpoint used by default to resolve the schema).

Reviewer 3: typo. most likely should be:

what will instruct the Schema Resolver `to`...

```java
Properties props = new Properties();

// Configure kafka settings
props.putIfAbsent(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, SERVERS);
props.putIfAbsent(ProducerConfig.CLIENT_ID_CONFIG, "Producer-" + TOPIC_NAME);
props.putIfAbsent(ProducerConfig.ACKS_CONFIG, "all");
props.putIfAbsent(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
props.putIfAbsent(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, JsonSchemaKafkaSerializer.class.getName());
props.putIfAbsent(SerdeConfig.ARTIFACT_RESOLVER_STRATEGY, SimpleTopicIdStrategy.class.getName());
props.putIfAbsent(SerdeConfig.EXPLICIT_ARTIFACT_GROUP_ID, "default");
props.putIfAbsent(SerdeConfig.REGISTRY_URL, REGISTRY_URL);

//This is the most important configuration for this particular example.

> props.putIfAbsent(SchemaResolverConfig.SERIALIZER_DEREFERENCE_SCHEMA, "true");

Reviewer 2: What does this do for the producer/serializer?

Author: In the case of JSON Schema, the schema is already registered, so when it's fetched to validate the data before it's being produced, it gets the schema dereferenced. For avro, if it's enabled and registration is enabled as well, it registers the schema dereferenced (since the schema is inferred from the data, not determined by the user).

    // Send/produce the message on the Kafka Producer
    ProducerRecord<Object, Object> producedRecord = new ProducerRecord<>(topicName, SUBJECT_NAME, citizen);
    producer.send(producedRecord);
}
```

And consume them with a consumer that is created as follows, instructing the deserializer to use the schema dereferenced with the following configuration property _DESERIALIZER_DEREFERENCE_SCHEMA_:

```java
Properties props = new Properties();

props.putIfAbsent(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, SERVERS);
props.putIfAbsent(ConsumerConfig.GROUP_ID_CONFIG, "Consumer-" + TOPIC_NAME);
props.putIfAbsent(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "true");
props.putIfAbsent(ConsumerConfig.AUTO_COMMIT_INTERVAL_MS_CONFIG, "1000");
props.putIfAbsent(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
props.putIfAbsent(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
props.putIfAbsent(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, JsonSchemaKafkaDeserializer.class.getName());
props.putIfAbsent(SerdeConfig.ARTIFACT_RESOLVER_STRATEGY, SimpleTopicIdStrategy.class.getName());

> props.putIfAbsent(SerdeConfig.EXPLICIT_ARTIFACT_GROUP_ID, "default");

Reviewer 2: Why is this used?

Reviewer 2: Nevermind I see.

props.putIfAbsent(SerdeConfig.VALIDATION_ENABLED, true);

//This is the most important configuration for this example, that instructs the deserializer to fetch the dereferenced schema.

> props.putIfAbsent(SchemaResolverConfig.DESERIALIZER_DEREFERENCE_SCHEMA, "true");

Reviewer 2: String as the value for this property rather than a boolean `true`?

Author: It "doesn't matter" in the sense that if the expected value is a boolean it's just parsed, but changing it to better honor the expected value.

// Configure Service Registry location
props.putIfAbsent(SerdeConfig.REGISTRY_URL, REGISTRY_URL);
// No other configuration needed for the deserializer, because the globalId of the schema
// the deserializer should use is sent as part of the payload.  So the deserializer simply
// extracts that globalId and uses it to look up the Schema from the registry.

// Create the Kafka Consumer
KafkaConsumer<Long, Citizen> consumer = new KafkaConsumer<>(props);
```

And then consume the messages with the consumer configured above like this:

```java
while (messageCount < 5) {
    final ConsumerRecords<Long, Citizen> records = consumer.poll(Duration.ofSeconds(1));
    messageCount += records.count();
    if (records.count() == 0) {
        // Do nothing - no messages waiting.
        System.out.println("No messages waiting...");
    } else records.forEach(record -> {
        Citizen msg = record.value();
        System.out.println("Consumed a message: " + msg + " @ " + msg.getCity());
    });
}
```

> With all of this, instead of doing an HTTP request to get the _citizen schema_, another one to fetch the pointers to the references, and, in this case (this process grows with the hierarchy size), another nine HTTP requests to discover the full hierarchy, a single HTTP request is made that gets the full content of the hierarchy and then uses it to produce/consume messages.

Reviewer: ```suggestion
Instead of doing an HTTP request to get the _citizen schema_, another request to fetch the pointers to the references, and, in this case (this process grows with the hierarchy size), another nine requests to discover the full hierarchy, a single HTTP request gets the full content of the hierarchy and then uses it to produce/consume messages.
```
